torchaudio/datasets/datasets.py

```diff
@@ -0,0 +1,419 @@
+class Cache:
+    """
+    Wrap a generator so that, whenever a new item is returned, it is saved to disk in a pickle.
+    """
+
+    def __init__(self, generator, location):
+        self.generator = generator
+        self.location = location
+
+        self._id = id(self)
+        self._cache = []
+        self._internal_index = 0
+
+    def __iter__(self):
+        self._internal_index = 0
+        return self
+
+    def __next__(self):
+        if self._internal_index < len(self):
+            item = self[self._internal_index]
+        else:
+            item = next(self.generator)
+
+            file = str(self._id) + "-" + str(len(self))
+            file = os.path.join(self.location, file)
+            self._cache.append(file)
+
+            os.makedirs(self.location, exist_ok=True)
+            with open(file, "wb") as file:
+                pickle.dump(item, file)
+
+        self._internal_index += 1
+        return item
+
+    def __getitem__(self, index):
+        file = self._cache[index]
+        with open(file, "rb") as file:
+            item = pickle.load(file)
+        return item
+
+    def __len__(self):
+        # Return length of cache
+        return len(self._cache)
+
+
+class Buffer:
+    """
+    Wrap a generator so as to keep the last few in memory.
+    """
+
+    def __init__(self, generator, capacity=10):
+        self.generator = generator
+        self.capacity = capacity
+        self._cache = []
+        self._fill()
+
+    def _fill(self):
+        while len(self._cache) <= self.capacity:
+            self._cache.append(next(self.generator))
+
+    def __getitem__(self, n):
+        self._fill()
+        return self._cache[n]
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        item = self._cache.pop(0)
+        self._fill()
+        return item
+
+
+def download(urls, root_path):
+    """
+    Download each url to root_path.
+
+    Input: url generator, folder inside archive
+    Output: downloaded archive, folder inside archive
+    """
+    for url, folder in urls:
+        torchaudio.datasets.utils.download_url(url, root_path)
+        file = os.path.join(root_path, os.path.basename(url))
+        yield file, folder
+
+
+def extract(files):
+    """
+    Extract each archive to their respective folder.
+
+    Input: (url, folder name inside archive) generator
+    Output: path to inside archive
+    """
+    for file, folder in files:
+        torchaudio.datasets.utils.extract_archive(file)
+        path = os.path.dirname(file)
+        path = os.path.join(path, folder)
+        yield path
+
+
+def walk(paths, extension):
+    """
+    Walk inside a path recursively to find all files with given extension.
+
+    Input: path
+    Output: path, file name identifying a row of data
+    """
+    for path in paths:
+        for dp, dn, fn in os.walk(path):
+            for f in fn:
+                if extension in f:
+                    yield path, f
+
+
+def shuffle(generator):
```

Comment:
I don't think this is necessary, except that random.shuffle is annoying because it doesn't give you a reference to the shuffled object.

Instead of generator you can also say "iterable"